Include the b2sum for the original data in the generated PDF file.

tuxlifan opened the pull request:
This aids in verification:
 1) after creation, that a PDF was created which can be successfully
    restored --> see new paperbackup-verify.sh script which in this role
    automates the diff step from the README.
 2) after restoration, by manually running b2sum on the restored data
    and comparing to the printout, that the restoration has indeed produced
    the exact same data as had been in the original file. This step could be
    automated as well in the future if we get a "metadata chunk".

Dependencies for paperbackup-verify.sh besides bash: coreutils, pdftotext (from poppler-utils), grep, diff

tuxlifan committed:
Include the b2sum for the original data in the generated PDF file.

This aids in verification
 1) after creation, that a PDF was created which can be successfully
    restored --> see new paperbackup-verify.sh script
 2) after restoration, by running b2sum on the restored data, that the
    restoration has indeed produced the exact same data as had been
    in the original file

intra2net commented:
Thanks for clarifying the need for the additional checksum in the discussion of Pull request #3.

Why do you use b2sum (BLAKE)? We just need the checksumming properties here, no crypto strength. I'd favor the much more common MD5 for that. MD5 has been around for a long time now and stable implementations are available on every platform. So even if blake is well standardized etc., I think md5sum has better availibility and will stick around in the future. CentOS/RHEL 7 for example does not have b2sum. I think the bit-length of md5 is good enough to protect the rather short paperbackup data against clashes.

So I'd prefer if we use md5 instead.

I'd also strip the "which()" and pathfinding code - I think when the user is able to call paperbackup.py in the first place, he should also be able to add ./ or the proper path to a call to paperbackup-verify.sh.

So I'd just output:
```
Please now verify that the output can be restored by calling:
paperbackup-verify.sh {}.pdf
```

tuxlifan committed:
Don't try to find paperbackup-verify.sh and simplify the message

tuxlifan committed:
New --b2sum option as alternative to b2sum binary

Usage:
  paperbackup.py --b2sum FILE
  echo "DATA" | paperbackup.py --b2sum

Print b2sum of FILE or DATA on stdout and exit.

NOTE: does not append the file name.

tuxlifan committed:
Switch to --b2sum in paperbackup.py i.s.o. b2sum binary

tuxlifan commented:
> Why do you use b2sum (BLAKE)?

By using a cryptographically strong hash we empower the user to copy it to a different backup location (which does only need to store that hash securely and not hold a full copy of the backup) and improve the tamper resistance of the backup scheme.

> CentOS/RHEL 7 for example does not have b2sum.

AFAICT python v3.6.0rc2 is the minimum required for blake2b and python36-3.6.3 is available via EPEL 7?
I've included a crude --b2sum option into paperbackup.py (until we switch to argparse to clean things up?) and use it instead of the b2sum call in paperbackup-verify.sh now.
Or am I missing the point?

> I'd also strip the "which()" and pathfinding code - I think when the user is able to call paperbackup.py in the first place, he should also be able to add ./ or the proper path to a call to paperbackup-verify.sh.

agreed

tuxlifan commented:
> So I'd prefer if we use md5 instead.

How would you feel to output both md5 and blake2b?

intra2net commented:
> By using a cryptographically strong hash we empower the user to copy it to a different 
> backup location (which does only need to store that hash securely and not hold a full copy 
> of the backup) and improve the tamper resistance of the backup scheme.

The objects which paperbackup was initially designed for - crypto keys and chiphertext - are usually already fully integrity protected by the crypto program, e.g. GnuPG, PKCS#8 or PKCS #12.

But if you store other kinds of data with paperbackup a cryptographically secure hash could help.

> How would you feel to output both md5 and blake2b?

How about SHA2-256 then? That is also well standardized, considered cryptographically secure and tools like sha256sum are widely available. The hex output still fits into one line on the pdf output.

BLAKE isn't part of the SHA family, so it's adoption and availbility could drop significantly in the future. Remember the history of e.g. Blowfish. It is still considered secure, but AES won the popularity vote by far. I recently saw the first programs removing support for Blowfish.

tuxlifan committed:
Switch from blake2b to sha256 checksum

Future proof by popularity.

tuxlifan commented:
Done :)

intra2net commented:
thank you for persevering
